FOCUS #252: Identifying custom columns (#252)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/attributes/column_naming_convention.md

> Column IDs provided in cost data following a consistent naming convention reduces friction for FinOps practitioners
> that consume the data for analysis, reporting, and other use cases.

Author: ```suggestion
Column IDs provided in cost data following a consistent naming convention reduces friction for FinOps practitioners that consume the data for analysis, reporting, and other use cases.
```

## Requirements

- All columns defined by FOCUS MUST follow the following rules:
  - Column IDs MUST use [Pascal case](https://techterms.com/definition/pascalcase).
  - Column IDs MUST NOT use abbreviations.
  - Column IDs SHOULD NOT use acronyms.
  - Column IDs MUST be alphanumeric with no special characters.
  - Columns that have an ID and a Name MUST have the `Id` or `Name` suffix in the Column ID. Display Name for a Column MAY avoid the `Name` suffix if there are no other columns with the same name prefix.
- Custom columns SHOULD follow the same rules as FOCUS columns listed above.

> - All custom column IDs MUST start with a vendor prefix string and an underscore (e.g., `abc_ColumnName`).

Reviewer: My preference would go to slash '/' or colon ': ' rather underscore '_', because underscore would make column names look like a mix of Snake case and Pascale case.
That is my first impression, but maybe I will get used to  '_'.
Curious to know what other think.

Author: The concern there is that most databases don't support them. Look at the supporting content. All the considerations are there. That said, I know slash will likely come in when we do grouping, so that's also something to consider.

Reviewer: Ok, but some column names of some databases like Oracle are case insensitive by default, while FOCUS uses Pascal case.
Moreover, AWS columns in CUR contain slash.
So I believe data ingestion would do some transformation anyway to cope with database naming constraints.
But Ok I understand the rationale, thanks!

  - Vendor prefixes MUST be lowercase, alphanumeric strings and SHOULD NOT be longer than 10 characters in length.

>   - Vendor prefixes MUST be representative of the company or brand that generated the column.

Reviewer: There is a discussion about how to avoid prefix conflicts between different vendors and a proposition here:
https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/264/files

Another possibility to avoid conflicts: vendors could use URL domain names as vendor prefix, for example: azr.io, aws.io, etc ... Vendors might need to buy domain names to get short prefixes.

What's also interesting with this:
- As a practitioner,  when I discover a prefix, I can go to the website to discover which vendor it is
- A a vendor, I can document my FOCUS column on the website

Author: Can you leave this feedback on that PR? I'd like to separate registration from at least having a prefix.

## Requirements

* All columns defined by FOCUS MUST follow the following rules:
  * Column IDs MUST use [Pascal case](https://techterms.com/definition/pascalcase).
  * Column IDs MUST NOT use abbreviations.
  * Column IDs SHOULD NOT use acronyms.
  * Column IDs MUST be alphanumeric with no special characters.
  * Columns that have an ID and a Name MUST have the `Id` or `Name` suffix in the Column ID. Display Name for a Column MAY avoid the Name suffix if there are no other columns with the same name prefix.

> * Custom columns SHOULD follow the same rules as FOCUS columns listed above.

Reviewer 2: The concept of a custom column is not defined anywhere. Would provider-defined column be a little more clearer?
```suggestion
* Provider-defined columns SHOULD follow the same rules as FOCUS columns listed above.
```

Reviewer 2: Sorry, trying to understand if this was addressed somehow or if there's a explanation for why we think 'Custom columns' is clear..

Author: Custom columns will come from many sources, not just "providers".

> * All custom columns MUST be prefixed with an underscore (`_`) to distinguish them from FOCUS columns and avoid conflicts in future releases.

Reviewer 2: ```suggestion
* All provider-defined columns MUST be prefixed with an underscore (`_`) to distinguish them from FOCUS columns and avoid conflicts in future releases.
```

Reviewer 3: Columns starting with '_' might require special treatment by different DB/query engines hence cause unnecessary issues for practitioners. For example '_' needs to be encapsulated in Amazon Athena and it's one of the commonly used services for querying cost and usage data on AWS (https://docs.aws.amazon.com/athena/latest/ug/tables-databases-columns-names.html#names-that-begin-with-an-underscore). My opinion that using pattern 'provider_customname' would help to avoid this. 
I understand that you won't be able to group all custom columns by sorting, but at the same time all modern IDEs or query editors have search filters which allow quickly search for columns from particular provider by name

> * All custom columns MUST be prefixed with an underscore (`_`) to distinguish them from FOCUS columns and avoid conflicts in future releases.

Author: ```suggestion
* All custom columns MUST be prefixed with a consistent `x_` prefix to identify them as external, custom columns and distinguish them from FOCUS columns to avoid conflicts in future releases.
```

> * All FOCUS columns MUST be first in the provided dataset.

Reviewer 4: Why are we being particular about the order? This seems odd.

Author: Usability. This has come up several times (even in 0.5).

Author: ```suggestion
* All FOCUS columns SHOULD be first in the provided dataset.
```

>   * Custom columns MUST be listed after all FOCUS columns and MUST NOT be intermixed.

Reviewer 2: ```suggestion
  * Provider-defined columns MUST be listed after all FOCUS columns and MUST NOT be intermixed.
```

> * All FOCUS columns MUST be first in provided dataset.
>   * Custom columns MUST be listed after all FOCUS columns and MUST NOT be intermixed.
>   * FOCUS and custom columns MAY be sorted alphabetically, but MUST remain separate.

Author: ```suggestion
* All FOCUS columns MUST be first in the provided dataset.
  * Custom columns MUST be listed after all FOCUS columns and MUST NOT be intermixed.
  * Columns MAY be sorted alphabetically but custom columns MUST be after all FOCUS columns.
```

>   * Columns MAY be sorted alphabetically but custom columns MUST be after all FOCUS columns.

Reviewer 2: ```suggestion
  * Columns MAY be sorted alphabetically but provider-defined columns MUST be after all FOCUS columns.
```

>   * Custom columns MUST be listed after all FOCUS columns and MUST NOT be intermixed.
>   * Columns MAY be sorted alphabetically but custom columns MUST be after all FOCUS columns.

Author: ```suggestion
  * Custom columns SHOULD be listed after all FOCUS columns and SHOULD NOT be intermixed.
  * Columns MAY be sorted alphabetically but custom columns SHOULD be after all FOCUS columns.
```

## Requirements

- All columns defined by FOCUS MUST follow the following rules:
  - Column IDs MUST use [Pascal case](https://techterms.com/definition/pascalcase).
  - Column IDs MUST NOT use abbreviations.
  - Column IDs SHOULD NOT use acronyms.

>   - Column IDs MUST be alphanumeric with no special characters.

Reviewer 5: What does "special characters" mean?  The composition of special characters changes from one platform to the next, and many such lists include underscores.  The spec should define these or link to an appropriate list.

Author: This line is about FOCUS columns and already exists today. We're saying that we cannot use underscores today. The one exception that's part of this PR is a vendor prefix for custom columns. If you think we should change it, can you start a thread in Slack or maybe just a separate PR? Not pushing back. Just don't want to add unrelated changes to this PR to avoid more delays.

## Requirements

- All columns defined by FOCUS MUST follow the following rules:
  - Column IDs MUST use [Pascal case](https://techterms.com/definition/pascalcase).

>   - Column IDs MUST NOT use abbreviations.

Reviewer 6: ```suggestion
  - Column IDs MUST NOT use abbreviations beyond exceptions listed within the Column Naming Convention guidelines. 
```

Author: Does this mean we need to add this caveat for every exception? I was assuming the exception was enough to call out that it's overriding the defined rules 🤔 Ultimately, I'm fine either way.

Author: Given this is separate from the prefix, I'm going to "won't fix" this one. Let's chat in Slack. If you feel strongly, I'm happy to push out a separate PR.

> - All columns defined by FOCUS MUST follow the following rules:
>   - Column IDs MUST use [Pascal case](https://techterms.com/definition/pascalcase).
>   - Column IDs MUST NOT use abbreviations.
>   - Column IDs SHOULD NOT use acronyms.
>   - Column IDs MUST be alphanumeric with no special characters.
>   - Columns that have an ID and a Name MUST have the `Id` or `Name` suffix in the Column ID. Display Name for a Column MAY avoid the Name suffix if there are no other columns with the same name prefix.
> - Custom columns SHOULD follow the same rules as FOCUS columns listed above.
> - All custom columns MUST be prefixed with an underscore (`_`) to distinguish them from FOCUS columns and avoid conflicts in future releases.

Author: ```suggestion
* All columns defined by FOCUS MUST follow the following rules:
  * Column IDs MUST use [Pascal case](https://techterms.com/definition/pascalcase).
  * Column IDs MUST NOT use abbreviations.
  * Column IDs SHOULD NOT use acronyms.
  * Column IDs MUST be alphanumeric with no special characters.
  * Columns that have an ID and a Name MUST have the `Id` or `Name` suffix in the Column ID. Display Name for a Column MAY avoid the Name suffix if there are no other columns with the same name prefix.
* Custom columns SHOULD follow the same rules as FOCUS columns listed above.
* All custom columns MUST be prefixed with an underscore (`_`) to distinguish them from FOCUS columns and avoid conflicts in future releases.
* All FOCUS columns MUST be first in provided dataset.
  * Custom columns MUST be listed after all FOCUS columns and MUST NOT be intermixed.
  * FOCUS and custom columns MAY be sorted alphabetically, but MUST remain separate.
```
